ouroboros-consensus-protocol/src/Ouroboros/Consensus/Protocol/Praos/VRF.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures             #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TypeApplications           #-}
+{-# OPTIONS_GHC -Wno-unticked-promoted-constructors #-}
+
+-- | This module implements VRF range extension as described in
+-- https://iohk.io/en/research/library/papers/on-uc-secure-range-extension-and-batch-verification-for-ecvrf/
+module Ouroboros.Consensus.Protocol.Praos.VRF (
+    InputVRF
+  , VRFUsage (..)
+  , mkInputVRF
+  , vrfLeaderCheckValue
+  , vrfNonceValue
+  ) where
+
+import           Cardano.Binary (ToCBOR)
+import           Cardano.Crypto.Hash (Blake2b_256, Hash, castHash, hashToBytes,
+                     hashWith)
+import qualified Cardano.Crypto.Hash as Hash
+import           Cardano.Crypto.Util
+                     (SignableRepresentation (getSignableRepresentation),
+                     bytesToNatural)
+import           Cardano.Crypto.VRF (OutputVRF (..), getOutputVRFBytes)
+import           Cardano.Ledger.BaseTypes (Nonce (NeutralNonce, Nonce))
+import           Cardano.Ledger.Crypto (Crypto (HASH, VRF))
+import           Cardano.Ledger.Serialization (runByteBuilder)
+import           Cardano.Ledger.Slot (SlotNo (SlotNo))
+import qualified Data.ByteString.Builder as BS
+import qualified Data.ByteString.Builder.Extra as BS
+import           GHC.Generics (Generic)
+import           NoThunks.Class (NoThunks)
+import           Numeric.Natural (Natural)
+
+-- | Input to the verifiable random function. Consists of the hash of the slot
+-- and the epoch nonce.
+newtype InputVRF = InputVRF {unUnifiedVRF :: Hash Blake2b_256 InputVRF}
+  deriving (Eq, Ord, Show, Generic)
+  deriving newtype (NoThunks, ToCBOR)
+
+instance SignableRepresentation InputVRF where
+  getSignableRepresentation (InputVRF x) = hashToBytes x
```

Comment:
It could, but it would rely on an orphan instance. While I'm not generally bothered by orphans in the right cases, I'd rather not rely on one here and be explicit.